### PR TITLE
SNS: Fixed the password-protected HTTPS endpoints autoconfirm

### PR DESCRIPTION
**Pull request by Ninir**
## Description
Hi everyone,

Just suggesting a fix for the SNS autoconfirm for HTTPs endpoints having a Basic authentication. 
The suggested change makes the user-defined endpoint obfuscating the password. 
Migrated from https://github.com/hashicorp/terraform/pull/9696

## Related issues:
- https://github.com/terraform-providers/terraform-provider-aws/issues/463

## TODOs
- [x] Errors If we can't find the subscription
- [x] Prefer the use of package net/url instead of regular expressions

### CODE REVIEW

**Comment by catsby** [comment]
Hey @Ninir how do we stand here, is this something you think you'll be able to wrap up?

**Comment by Ninir** [comment]
Hey @catsby 
I'll have more time starting from tomorrow, so will get this into shape.
Already have the local work, just need to check a secure endpoint, using a personal VM at the moment.

**Comment by Ninir** [comment]
Here it is!

@radeksimko To better explain why I didn't implement DiffSuppressFunc: 
when the check to find a valid subscription is made, we are calling `snsconn.ListSubscriptionsByTopic`, which returns this kind of response:
```xml
Response sns/ListSubscriptionsByTopic Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK

-----------------------------------------------------
<ListSubscriptionsByTopicResponse xmlns="http://sns.amazonaws.com/doc/2010-03-31/">
  <ListSubscriptionsByTopicResult>
    <Subscriptions>
      <member>
        <Owner>123456789643</Owner>
        <Endpoint>https://myuser:****@mydomain.com</Endpoint>
        <Protocol>https</Protocol>
        <SubscriptionArn>arn:aws:sns:eu-west-1:123456789643:terraform-test-topic:ed16f123-61e3-4cbe-a9er-hzv9fz8vh</SubscriptionArn>
        <TopicArn>arn:aws:sns:eu-west-1:123456789643:terraform-test-topic</TopicArn>
      </member>
    </Subscriptions>
  </ListSubscriptionsByTopicResult>
</ListSubscriptionsByTopicResponse>
```

When the Read function is then called, `snsconn.GetSubscriptionAttributes` is called, and returns such stuff:
```xml
Response sns/GetSubscriptionAttributes Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK

-----------------------------------------------------
<GetSubscriptionAttributesResponse xmlns="http://sns.amazonaws.com/doc/2010-03-31/">
  <GetSubscriptionAttributesResult>
    <Attributes>
      <entry>
        <key>Owner</key>
        <value>123456789643</value>
      </entry>
      <entry>
        <key>RawMessageDelivery</key>
        <value>false</value>
      </entry>
      <entry>
        <key>TopicArn</key>
        <value>arn:aws:sns:eu-west-1:123456789643:terraform-test-topic</value>
      </entry>
      <entry>
        <key>Endpoint</key>
        <value>https://myuser:mypassword@mydomain.com</value>
      </entry>
      <entry>
        <key>EffectiveDeliveryPolicy</key>
        <value>{&quot;healthyRetryPolicy&quot;:{&quot;minDelayTarget&quot;:20,&quot;maxDelayTarget&quot;:20,&quot;numRetries&quot;:3,&quot;numMaxDelayRetries&quot;:0,&quot;numNoDelayRetries&quot;:0,&quot;numMinDelayRetries&quot;:0,&quot;backoffFunction&quot;:&quot;linear&quot;},&quot;sicklyRetryPolicy&quot;:null,&quot;throttlePolicy&quot;:null,&quot;guaranteed&quot;:false}</value>
      </entry>
      <entry>
        <key>Protocol</key>
        <value>https</value>
      </entry>
      <entry>
        <key>ConfirmationWasAuthenticated</key>
        <value>false</value>
      </entry>
      <entry>
        <key>SubscriptionArn</key>
        <value>arn:aws:sns:eu-west-1:123456789643:terraform-test-topic:ed16f123-61e3-4cbe-a9er-hzv9fz8vh</value>
      </entry>
    </Attributes>
  </GetSubscriptionAttributesResult>
</GetSubscriptionAttributesResponse>
```

In the end, the endpoint stored in the state is the correct one, and no diff is exposed.
The obfuscated endpoint thing is just happening when getting subscriptions, which match the console behaviour.

**Comment by radeksimko** [comment]
> Thus, I think we could probably set up a testacc using CloudFront pointing to an ELB with a VM...
Thus, we could use Radek's code, and check the auth + subscription (a lot of stuff there).

Can't we just use the API Gateway + Lambda? Booting a VM for this purpose feels a bit heavy.

> To better explain why I didn't implement DiffSuppressFunc:
> when the check to find a valid subscription is made, we are calling snsconn.ListSubscriptionsByTopic, which returns this kind of response

I see, so we only do this `***` masking to verify we have/haven't already subscribed to the topic because `GetSubscriptionAttributes` doesn't provide such information. That's 👌 then.

**Comment by Ninir** [comment]
@radeksimko Well, I tried to build a Basic-protected API Gateway using API Gateway + a custom Authorizer (Lambda). Everything works as expected using the cli, but I can't get the subscription to pass it. Not sure what is issuing here...

If you want, we can discuss it, and I can provide all the work i've made.
Is the test required for this to pass?

**Comment by radeksimko** [comment]
@Ninir I built a few API Gateway integrations in the past using Terraform, none of them used custom authorizer, but I'm confident we can build it.

Just ping me on Slack and/or share the config with me when you have a moment, I'm happy to pair up on this.

**Comment by Ninir** [comment]
Using API GW+Lambda+SNS is the easiest way to test things, indeed.

As discussed on Slack, it was not possible to set-up an acc test until we could specify headers on responses returned by API Gw.
Now that the API GW Gateway Response resource has been developed & integrated, i was able to set a `WWW-Authenticate` header on the `401` response returned by API Gateway when having a custom authorizer.

Here is what SNS does when requesting a HTTPS password-protected endpoint, like `https://john:doe@mydomain.com/myroute`:
1. Send a request to `https://mydomain.com/myroute` (without credentials)
2. If the endpoint returns a `WWW-Authenticate: Basic` header, SNS then sends a second request to the endpoint WITH credentials

The actual code is written in NodeJS, but could probably be rewritten in Python. @radeksimko Does it really matter?

File: aws/resource_aws_sns_topic_subscription_test.go [hunk 1]
```diff
@@ -245,3 +279,204 @@ resource "aws_sns_topic_subscription" "test_subscription" {
...
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_api_gateway_authorizer" "test" {
+  name                             = "tf-acc-test-api-gw-authorizer-%d"
+  rest_api_id                      = "${aws_api_gateway_rest_api.test.id}"
+  authorizer_uri                   = "arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/${aws_lambda_function.authorizer.arn}/invocations"
```
Comment (radeksimko) [inline aws/resource_aws_sns_topic_subscription_test.go]: Do you mind changing this to

```
${aws_lambda_function.authorizer.invoke_arn}
```
? That will make it a bit easier to read and (more importantly) make it work in `us-west-2` region which we use for acceptance tests 😉

**Comment by radeksimko** [comment]
> The actual code is written in NodeJS, but could probably be rewritten in Python. @radeksimko Does it really matter?

It doesn't matter at all here in this context, I think.

**Comment by Ninir** [comment]
Guh... @radeksimko Need holidays! 😄 

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSNSTopicSubscription_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSSNSTopicSubscription_ -timeout 120m
=== RUN   TestAccAWSSNSTopicSubscription_importBasic
--- PASS: TestAccAWSSNSTopicSubscription_importBasic (17.10s)
=== RUN   TestAccAWSSNSTopicSubscription_basic
--- PASS: TestAccAWSSNSTopicSubscription_basic (14.75s)
=== RUN   TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint (46.41s)
=== RUN   TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint (67.26s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	67.292s
```

**Comment by radeksimko** [comment]
No worries, thanks for all the effort in taking it to the finish line. 👍  😃

**Pull request merged**